File: gym/blade/mission/PatrolMission.py
import json
from typing import List
from random import random
from shapely.geometry import Point, Polygon
from blade.units.ReferencePoint import ReferencePoint


class PatrolMission:

    def __init__(
        self,
        id: str,
        name: str,
        side_id: str,
        assigned_unit_ids: List[str],
        assigned_area: List[ReferencePoint],
        active: bool,
    ):
        self.id = id
        self.name = name
        self.side_id = side_id
        self.assigned_unit_ids = assigned_unit_ids
        self.assigned_area = assigned_area
        self.active = active
        self.patrol_area_geometry = Polygon(
            [(point.longitude, point.latitude) for point in self.assigned_area]
        )

    def update_patrol_area_geometry(self):
        self.patrol_area_geometry = Polygon(
            [(point.longitude, point.latitude) for point in self.assigned_area]
        )

    def check_if_coordinates_is_within_patrol_area(
        self, coordinates: List[float]
    ) -> bool:
        point = Point(coordinates)
        if self.patrol_area_geometry.contains(point):
            return True
        return False

    def generate_random_coordinates_within_patrol_area(self) -> List[float]:
        random_coordinates = [
            random() * (self.assigned_area[2].latitude - self.assigned_area[0].latitude)
            + self.assigned_area[0].latitude,
            random()
            * (self.assigned_area[1].longitude - self.assigned_area[0].longitude)
            + self.assigned_area[0].longitude,
        ]
        return random_coordinates

    def toJSON(self):
        return json.dumps(
            self,
            default=lambda o: o.__dict__ if hasattr(o, "__dict__") else "",
            sort_keys=True,
            indent=4,
        )

File: client/src/game/mission/PatrolMission.ts
import { fromLonLat, get as getProjection } from "ol/proj";
import { DEFAULT_OL_PROJECTION_CODE } from "@/utils/constants";
import { Polygon } from "ol/geom";
import ReferencePoint from "@/game/units/ReferencePoint";

interface IPatrolMission {
  id: string;
  name: string;
  sideId: string;
  assignedUnitIds: string[];
  assignedArea: ReferencePoint[];
  active: boolean;
}

export default class PatrolMission {
  id: string;
  name: string;
  sideId: string;
  assignedUnitIds: string[];
  assignedArea: ReferencePoint[];
  active: boolean;
  patrolAreaGeometry: Polygon;

  constructor(parameters: IPatrolMission) {
    this.id = parameters.id;
    this.name = parameters.name;
    this.sideId = parameters.sideId;
    this.assignedUnitIds = parameters.assignedUnitIds;
    this.assignedArea = parameters.assignedArea;
    this.active = parameters.active;
    this.patrolAreaGeometry = this.createPatrolAreaGeometry(
      parameters.assignedArea
    );
  }

  updatePatrolAreaGeometry(): void {
    this.patrolAreaGeometry = this.createPatrolAreaGeometry(this.assignedArea);
  }

  createPatrolAreaGeometry(area: ReferencePoint[]): Polygon {
    const projection = getProjection(DEFAULT_OL_PROJECTION_CODE);
    return new Polygon([
      area.map((point) =>
        fromLonLat([point.longitude, point.latitude], projection!!)
      ),
    ]);
  }

  checkIfCoordinatesIsWithinPatrolArea(coordinates: number[]): boolean {
    const projection = getProjection(DEFAULT_OL_PROJECTION_CODE);
    return this.patrolAreaGeometry.intersectsCoordinate(
      fromLonLat([coordinates[1], coordinates[0]], projection!!)
    );
  }

  generateRandomCoordinatesWithinPatrolArea(): number[] {
    const randomCoordinates = [
      Math.random() *
        (this.assignedArea[2].latitude - this.assignedArea[0].latitude) +
        this.assignedArea[0].latitude,
      Math.random() *
        (this.assignedArea[1].longitude - this.assignedArea[0].longitude) +
        this.assignedArea[0].longitude,
    ];
    return randomCoordinates;
  }
}
